### PR TITLE
cz/cluster transform and SPRUCE support

### PR DESCRIPTION
**Pull request by ChuanyiZ**
- Add a step for transforming the output format of pyclone and pyclone-vi into SPRUCE's input. Currently only support converting "data file", not "interval file".
- Add a step for SPRUCE: including a Docker image and corresponding WDL scripts for SPRUCE. Only takes "data file" as input, not included "interval file" yet.
- Add test data files in folder `example_data/cluster`:
    - `example_data.tsv`: toy example input for SPRUCE, should produce 3 solutions.
    - `example_data_AML.tsv`: input file for SPRUCE generated by pyclone-vi on the AML dataset.
    - `pyclone_vi_AML_cluster_assignment.tsv`: pyclone-vi clustering output of the AML dataset, required by `cluster_transform`.
    - `pyclone_vi_AML_input.tsv`: pyclone-vi input of the AML dataset, required by `cluster_transform`.
    - `pyclone_vi_cluster_assignment.tsv`: pyclone-vi clustering output of the mutect2 sample (A25), required by `cluster_transform`.
    - `pyclone_vi_input.tsv`: pyclone-vi input of the mutect2 sample (A25), required by `cluster_transform`.
    - `pyclone_cluster_assignment.tsv`: pyclone clustering output of the mutect2 sample (A25), required by `cluster_transform`.
- Using SPRUCE commit [e5ffc8](https://github.com/raphael-group/spruce/commit/e5ffc8882fd4ea92bac9462a16ab81834c4f58ed)

### CODE REVIEW

File: cluster_transform/code/py_code/main.py [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+                                        vaf_lb,
+                                        vaf_mean,
+                                        vaf_ub))
+    n_cluster = len(df_clusters["cluster_id"].unique())
+    n_sample = len(sample_to_id)
+    return list_clustered, n_cluster, n_sample
+
+
+def write_spruce(list_clustered, n_cluster, n_sample, tsv_file):
+    with open(tsv_file, 'w') as ofile:
```
Comment (pgroves) [inline cluster_transform/code/py_code/main.py:103]: No need to change it here, but FYI there is a csv.writer class that can handle the details of adding the separators and newlines for you (for every line you give it a dict where the keys are column names).
https://docs.python.org/3/library/csv.html

File: cluster_transform/docker/Dockerfile-cluster-transform [hunk 2]
```diff
@@ -0,0 +1,82 @@
+# This container installs pyVCF and the code in phyloflow/vcf_transform/code
```
Comment (pgroves) [inline cluster_transform/docker/Dockerfile-cluster-transform:1]: The above line needs updating

File: spruce/docker/Dockerfile-spruce [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+    mkdir build && cd build && \
+    cmake .. && make && make install
+
+#####
+##Above this line could be a base image
+######
+
+# Install SPRUCE
+
+RUN git clone https://github.com/raphael-group/spruce.git && \
```
Comment (pgroves) [inline spruce/docker/Dockerfile-spruce:34]: We should probably be specifying at least the branch for the git repos when we do these "clone then build" operations. I will make an issue for later.

**Review by pgroves — APPROVED** [review]
Looks good. Feel free to merge when you are happy with it. I made a few comments on documentation and comments, but those don't really need to be perfect during this phase of the project. It will all need to be cleaned up if the larger project goes forward.

File: cluster_transform/README.md [hunk 4]
```diff
@@ -0,0 +1,3 @@
+# Cluster Transform Task
+
+WDL task to input cluster information and output ...
```
[the remark this answers is not in the document]
Comment (ChuanyiZ) [reply cluster_transform/README.md:3]: 👌

File: cluster_transform/code/py_code/main.py [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+                                        vaf_ub))
+    n_cluster = len(df_clusters["cluster_id"].unique())
+    n_sample = len(sample_to_id)
+    return list_clustered, n_cluster, n_sample
+
+
+def get_cluster_pyclone_vi(cluster_file, tsv_files):
+    df_clusters = pd.read_csv(cluster_file, sep='\t',
+                              dtype={"mutation_id": bytes, "sample_id": bytes})
+    df_clusters["mutation_id"] = df_clusters["mutation_id"].apply(lambda x: ast.literal_eval(x).decode("utf-8"))
```
[the remark this answers is not in the document]
Comment (ChuanyiZ) [reply cluster_transform/code/py_code/main.py:77]: 👌

File: cluster_transform/code/py_code/main.py (hunk 1 above)
Comment (ChuanyiZ) [reply cluster_transform/code/py_code/main.py:103]: Thanks.

File: cluster_transform/docker/Dockerfile-cluster-transform (hunk 2 above)
Comment (ChuanyiZ) [reply cluster_transform/docker/Dockerfile-cluster-transform:1]: 👌

File: spruce/docker/Dockerfile-spruce (hunk 3 above)
Comment (ChuanyiZ) [reply spruce/docker/Dockerfile-spruce:34]: Sounds good.

**Pull request merged**